service-workers/service-worker/ServiceWorkerGlobalScope/extendable-message-event.https.html

```diff
@@ -5,222 +5,221 @@
-              constructor: { name: 'ExtendableMessageEvent' },
-              origin: location.origin,
-              lastEventId: '',
-              source: {
-                  constructor: { name: 'ServiceWorker' },
-                  scriptURL: normalizeURL(script2),
-                  state: 'installed'
-              },
-              ports: []
-          };
-          assert_equals(results[1].pingOrPong, 'pong');
-          ExtendableMessageEventUtils.assert_equals(
-              results[1].event, expected_pong
-          );
-        });
-  }, 'Post extendable messages among service workers');
+promise_test(async t => {
+  const SCRIPT = 'resources/extendable-message-event-worker.js';
+  const SCOPE = 'resources/scope/extendable-message-event-from-toplevel';
+  const registration = await service_worker_unregister_and_register(
+    t,
+    SCRIPT,
+    SCOPE
+  );
+
+  t.add_cleanup(async() => {
+    if (registration) await registration.unregister();
+  });
+
+  await wait_for_state(t, registration.installing, 'activated');
+  const saw_message = new Promise(function(resolve) {
+    navigator.serviceWorker.onmessage = function(event) { 
+      resolve(event.data);
+    };
+  });
+  const channel = new MessageChannel;
+  registration.active.postMessage('', [channel.port1]);
+  const results = await saw_message;
+  const expected = {
+    constructor: { name: 'ExtendableMessageEvent' },
+    origin: location.origin,
+    lastEventId: '',
+    source: {
+      constructor: { name: 'WindowClient' },
+      frameType: 'top-level',
+      url: location.href,
+      visibilityState: 'visible',
+      focused: true
+    },
+    ports: [ { constructor: { name: 'MessagePort' } } ]
+  };
+  ExtendableMessageEventUtils.assert_equals(results, expected);
+}, 'Post an extendable message from a top-level client');
+
+promise_test(async t => {
+  const SCRIPT = 'resources/extendable-message-event-worker.js';
+  const SCOPE = 'resources/scope/extendable-message-event-from-nested';
+
+  const registration = await service_worker_unregister_and_register(
+    t,
+    SCRIPT,
+    SCOPE
+  );
+
+  await wait_for_state(t, registration.installing, 'activated');
+  const frame = await with_iframe(SCOPE);
+  const saw_message = new Promise(function(resolve) {
+    frame.contentWindow.navigator.serviceWorker.onmessage = function(event) {
+      resolve(event.data);
+    };
+  });
+  frame.contentWindow.navigator.serviceWorker.controller.postMessage('');
+  const results = await saw_message;
+  const expected = {
+    constructor: { name: 'ExtendableMessageEvent' },
+    origin: location.origin,
+    lastEventId: '',
+    source: {
+      constructor: { name: 'WindowClient' },
+      url: frame.contentWindow.location.href,
+      frameType: 'nested',
+      visibilityState: 'visible',
+      focused: false
+    },
+    ports: []
+  };
+
+  t.add_cleanup(async() => {
+    if (frame) frame.remove();
+    if (registration) await registration.unregister();
+  });
```

Comment:
I try to keep the add_cleanup() near the top of the test so it gets run if any await fails. It might require making |frame| and |registration| non-const so they can be defined before the add_cleanup() call.

Reply:
Removed the add_cleanup()  to near the top of the test.